onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,368 @@
+      auto& n_input = graph_build.GetOrCreateNodeArg(input->Name(), input->TypeAsProto());
+      inputs.push_back(&n_input);
+      all_node_inputs.insert(input->Name());
+    }
+    for (auto output : node.OutputDefs()) {
+      auto& n_output = graph_build.GetOrCreateNodeArg(output->Name(), output->TypeAsProto());
+      outputs.push_back(&n_output);
+    }
+    graph_build.AddNode(node.Name(), node.OpType(), node.Description(), inputs, outputs, &node.GetAttributes(), node.Domain());
+  }
+  //Add initializer to graph
+  const auto& init_tensors = graph_view.GetAllInitializedTensors();
+  for (const auto& tensor : init_tensors) {
+    graph_build.AddInitializedTensor(*(tensor.second));
+  }
+
+  ORT_ENFORCE(graph_build.Resolve().IsOK());
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  nnapi::ModelBuilder builder(model_proto);
+  const auto supported_nodes_vector = builder.GetSupportedNodes();
+
+  // Find inputs, initializers and outputs for each supported subgraph
+  const std::vector<NodeIndex>& node_index = graph_view.GetNodesInTopologicalOrder();
+  const auto graph_outputs = graph_view.GetOutputs();
+  int counter = 0;
+  for (const auto& group : supported_nodes_vector) {
+    if (group.empty())
+      continue;
+
+    std::unordered_set<size_t> node_set;
+    node_set.reserve(group.size());
+    for (const auto& index : group) {
+      node_set.insert(node_index[index]);
+    }
+
+    std::unique_ptr<IndexedSubGraph> sub_graph = onnxruntime::make_unique<IndexedSubGraph>();
+    // Find inputs and outputs of the subgraph
+    std::unordered_map<const NodeArg*, int> fused_inputs, fused_outputs, fused_outputs_to_add;
+    std::unordered_set<const NodeArg*> erased;
+    int input_order = 0;
+    int output_order = 0;
+
+    for (const auto& index : group) {
+      sub_graph->nodes.push_back(node_index[index]);
+      const auto& node = graph_view.GetNode(node_index[index]);
+
+      for (const auto& input : node->InputDefs()) {
+        const auto& it = fused_outputs.find(input);
+        if (it != fused_outputs.end()) {
+          fused_outputs.erase(it);
+          erased.insert(input);
+        }
+        //only when input is neither in output list nor erased list, add the input to input list
+        else if (erased.find(input) == erased.end()) {
+          fused_inputs[input] = input_order++;
+        }
+      }
+
+      // For output searching, there is a special case:
+      // If certain output is used more than once,
+      // if the output is connected to nodes that don't belong to the subgraph, the output need to be added
+      // to the output list
+
+      std::unordered_set<const NodeArg*> processed_outputs;
+      for (auto it = node->OutputEdgesBegin(), end = node->OutputEdgesEnd(); it != end; ++it) {
+        const auto& node_idx = it->GetNode().Index();
+        const auto& output = (it->GetNode()).InputDefs()[it->GetDstArgIndex()];
+
```

Comment:
This returns const NodeArg* doesn't it, so 'const auto&' is confusing?

I would have expected it's the same value as node->OutputDefs()[it->GetSrcArgIndex()] and that may be slightly easier to understand